---
title: "Drug Targets by Pathway"
keep_md: no
---
```{r Read files and prepare framework for analyis}
library(pacman)

p_load(data.table, edgeR, pheatmap, viridis, tidyverse, hrbrthemes, poolr)

# reads
aba.counts <- fread(
	"../../all_counts_seal.tsv.gz",
	header = FALSE,
	col.names = c("spacer", "count", "condition"))

# guide design key
aba.key <- fread("../../aba_key.tsv")

# experimental design
aba.design <- fread("../../ABA1_experimental_design.tsv")

# hand-curated names with references available
curated_names <- fread("../../curated_names.tsv")

# Pathway definitions
# Cell wall biogenesis/degradation, and Cell Wall/PG
CellWall <- fread("../../CL704.tsv") 

# Ribosome, why is rplY and rpmB not being painted in plots?
map03010 <- fread("../../map03010.tsv") 

# Lipopolysaccharide Genes
LOS <- fread("../../LOS.tsv") 

# NADH Genes
NADH <- fread("../../NADH.tsv") 

# Aminoacyl-tRNA synthetase, GO
GO0004812 <- fread("../../GO0004812.tsv")

# https://version-11-5.string-db.org/cgi/globalenrichment?networkId=b1Fd534fBhzJ
CL707 <- fread ("../../CL707.tsv") # "Cell Shape/Lipid Glyc."
CL273 <- fread ("../../CL273.tsv") # "tRNA Aminoacylation"
```

```{r Functions for use in results parsing}

extract_edgeR = function(y.fit, y.contrasts, y.permutations) {
	
	contrasts.edgeR <- makeContrasts(
		contrasts = y.contrasts,
		levels = y.permutations)

	results <- glmQLFTest(y.fit, contrast = contrasts.edgeR)
	
	results <- topTags(results, n = Inf)
	
	results <- data.table(results$table)
	
	return(results)
}

clean_edgeR = function(results) {
	
	col_rep <- "\\.\\.\\." #i.e. "..." -> " - " 
	
	setnames(
		results, 
		names(results[, .SD, .SDcols = patterns(col_rep)]), 
		gsub(col_rep, " - ",  names(results[, .SD, .SDcols = patterns(col_rep)])))
	
	col_rep <- "logFC\\." #i.e. "logFC." -> ""
	
	setnames(
		results, 
		names(results[, .SD, .SDcols = patterns(col_rep)]), 
		gsub(col_rep, "",  names(results[, .SD, .SDcols = patterns(col_rep)])))
	
	return(results)
}

melt_and_stack = function(results) {
	
	results <- aba.genome[
		results, on = .(spacer == genes)]

	melted_results <- data.table::melt(
		results,
		id.vars = c(
			"AB19606",
			"AB030",
			"unique_name",
			"spacer",
			"type",
			"y_pred",
			"target",
			"offset",
			"FDR",
			"Pathway"),
		variable.name = "condition", 
		value.name = "LFC",
		measure.vars = aba.contrasts)
	
	melted_results <- melted_results[!is.na(FDR) & !is.na(LFC)]
	
	control_melted_results_by_condition <- 
		melted_results[
			type == "control",
			.(med_LFC = median(LFC)),
			keyby = .(condition)]
	
	setkey(melted_results, condition)
	
	melted_results[
		, LFC.adj := control_melted_results_by_condition[
			melted_results, LFC - med_LFC, by = .EACHI]$V1]
	
	return(melted_results)
}

```

```{r Functions for use in heatmaps}
generate_breaks = function(x, n, center = F) {
	
	if (center) {
		m = max(abs(c(min(x, na.rm = T), max(x, na.rm = T))))
		res = seq(-m, m, length.out = n + 1)}
	
	else {
		res = seq(min(x, na.rm = T), max(x, na.rm = T), length.out = n + 1)}
	
	return(res)
}

plot_colors <- c(
	colorRampPalette(
		c("#9a0007", "white"))(144^3) %>% 
		`[`((1:144)^3) %>% 
		`[`(-144),
	"white",
	rev(colorRampPalette(
		c("#005b9f", "white"))(144^3) %>% 
			`[`((1:144)^3) %>% 
			`[`(-144)))

```


```{r Combine, format, and filter data}
aba.genome <- aba.key[
	, .(spacer, type, locus_tag, y_pred, target, offset), by = .(locus_tag)]

aba.genome <- curated_names[
	, .(AB19606, AB030, unique_name)][aba.genome, on = .(AB19606 == locus_tag)]

# classify genes as belonging to one pathway
aba.genome[AB030 %in% CellWall$AB030, Pathway := "Cell Wall/PG"]
aba.genome[AB030 %in% map03010$AB030, Pathway := "Ribosome"]
aba.genome[AB030 %in% LOS$AB030, Pathway := "LOS"]
aba.genome[AB030 %in% NADH$AB030, Pathway := "NADH"]
aba.genome[AB030 %in% GO0004812$AB030, Pathway := "tRNA Ligase"]
aba.genome[is.na(Pathway), Pathway := "Other"]
aba.genome[AB19606 == "", AB19606 := NA]

# define the experimental design space to only take into consideration "tubes"
aba.design <- aba.design[experiment == "tube"]

# keep only the counts that are in the experimental design space
aba.counts <- aba.counts[condition %in% aba.design$condition]

# convert single column into a table 
aba.grid <- data.table::dcast(
	aba.counts, 
	spacer ~ factor(condition, levels = unique(condition)),
	value.var = "count", 
	fill = 0)

aba.grid.mat <- 
	data.matrix(aba.grid[, -c("spacer")])

row.names(aba.grid.mat) <- aba.grid$spacer

aba.group <- factor(aba.design[, paste(drug, dose, timing, sep = "_")])

aba.permutations <-	model.matrix( ~ 0 + aba.group)

colnames(aba.permutations) <- levels(aba.group)
```

```{r Prepare data for edgeR}
aba.y <- DGEList(
	counts = aba.grid.mat,
	group = aba.group,
	genes = row.names(aba.grid.mat))

aba.keep <- filterByExpr(
	y = aba.y,
	design = aba.permutations,
	group = aba.group)

aba.y <- aba.y[aba.keep, , keep.lib.sizes = TRUE]

aba.y <- calcNormFactors(aba.y)

aba.y <- estimateDisp(aba.y, aba.permutations)

aba.fit <- glmQLFit(aba.y, aba.permutations, robust = TRUE)

```

```{r Perform edgeR for induced-only samples}
aba.contrasts <- c(
	"None_0_T1 - None_0_T0",
	"None_0_T2 - None_0_T0",
	"None_0_T2 - None_0_T1") 

results <- extract_edgeR(
	y.fit = aba.fit, 
	y.contrasts = aba.contrasts, 
	y.permutations = aba.permutations)

results <- clean_edgeR(results)

melted_results <- melt_and_stack(results)
```



```{r nuo activity in induced samples, fig.width=10, fig.height=10}
median_LFC <- dcast(
	melted_results[Pathway == "NADH" & type == "perfect"], 
	unique_name + Pathway ~ condition, 
	value.var = "LFC.adj", 
	fun.aggregate = median)

print(colnames(median_LFC))

to_plot_title <- "Median LogFC in Perfect Guides"

# convert to matrix
LFC_grid <- data.matrix(median_LFC[, c(-1:-2)])

# name rows in matrix
rownames(LFC_grid) <- median_LFC[, paste(unique_name)]

# format genes (rows): Pathway + italic(Gene Name)
genes.formatted <- lapply(
	median_LFC[, .I],
	function(x) { 
		bquote(.(median_LFC[x]$Pathway) ~ italic(.(median_LFC[x]$unique_name)))})

# format conditions (columns)
conditions.formatted <- c(
	bquote(Induction ~ (t[1] - t[0])),
	bquote(Induction ~ (t[2] - t[0])),
	bquote(Difference ~ (t[2] - t[1])))

plot_matrix <- LFC_grid

breaks <- generate_breaks(plot_matrix, n = 286, center = T)

to_plot <- pheatmap(
	plot_matrix,
	col = plot_colors,
	breaks = breaks,
	border_color = NA,
	main = to_plot_title,
	cluster_cols = F,
	angle_col = 45,
	clustering_method = "ward.D2",
	clustering_distance_rows = "canberra",
	labels_row = as.expression(genes.formatted),
	labels_col = as.expression(conditions.formatted),
	# gaps_col = 3,
	# cutree_rows = 4,
	cellheight = 24,
	cellwidth = 48,
	fontsize = 12,
	silent = T)
	
to_plot$gtable$grobs[[1]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[1]]$gp$fontsize <- 18
to_plot$gtable$grobs[[4]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[5]]$gp$fontfamily <- "Arial"
	
print(to_plot)
```

```{r Perform edgeR for Colistin Samples}
aba.contrasts <- c(
	"None_0_T1 - None_0_T0",
	"Colistin_0.44_T1 - None_0_T0",
	"Colistin_0.44_T1 - None_0_T1",
	"None_0_T2 - None_0_T0",
	"Colistin_0.44_T2 - None_0_T0",
	"Colistin_0.44_T2 - None_0_T2") 

results <- extract_edgeR(
	y.fit = aba.fit, 
	y.contrasts = aba.contrasts, 
	y.permutations = aba.permutations)

results <- clean_edgeR(results)

melted_results <- melt_and_stack(results)
```


```{r LOS activity in colistin, fig.width=10, fig.height=10}
median_LFC <- dcast(
	melted_results[Pathway == "LOS" & type == "perfect"], 
	unique_name + Pathway ~ condition, 
	value.var = "LFC.adj", 
	fun.aggregate = median)

print(colnames(median_LFC))

to_plot_title <- "Median LogFC in Perfect Guides"

# convert to matrix
LFC_grid <- data.matrix(median_LFC[, c(-1:-2)])

# name rows in matrix
rownames(LFC_grid) <- median_LFC[, paste(unique_name)]

# format genes (rows): Pathway + italic(Gene Name)
genes.formatted <- lapply(
	median_LFC[, .I],
	function(x) { 
		bquote(.(median_LFC[x]$Pathway) ~ italic(.(median_LFC[x]$unique_name)))})

# format conditions (columns)
conditions.formatted <- c(
	bquote(Induction ~ (t[1] - t[0])),
	bquote(Colistin ~ (t[1] - t[0])),
	bquote(Colistin - Induction ~ (t[1] - t[1])),
	bquote(Induction ~ (t[2] - t[0])),
	bquote(Colistin ~ (t[2] - t[0])),
	bquote(Colistin - Induction ~ (t[2] - t[2])))

plot_matrix <- LFC_grid

breaks <- generate_breaks(plot_matrix, n = 286, center = T)

to_plot <- pheatmap(
	plot_matrix,
	col = plot_colors,
	breaks = breaks,
	border_color = NA,
	main = to_plot_title,
	cluster_cols = F,
	angle_col = 45,
	clustering_method = "ward.D2",
	clustering_distance_rows = "canberra",
	labels_row = as.expression(genes.formatted),
	labels_col = as.expression(conditions.formatted),
	gaps_col = 3,
	cutree_rows = 4,
	cellheight = 24,
	cellwidth = 48,
	fontsize = 12,
	silent = T)
	
to_plot$gtable$grobs[[1]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[1]]$gp$fontsize <- 18
to_plot$gtable$grobs[[4]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[5]]$gp$fontfamily <- "Arial"
	
print(to_plot)
```

```{r Perform edgeR for Rifampicin Samples}
aba.contrasts <- c(
	"None_0_T1 - None_0_T0",
	"Rifampicin_0.34_T1 - None_0_T0",
	"Rifampicin_0.34_T1 - None_0_T1",
	"None_0_T2 - None_0_T0",
	"Rifampicin_0.34_T2 - None_0_T0",
	"Rifampicin_0.34_T2 - None_0_T2") 

results <- extract_edgeR(
	y.fit = aba.fit, 
	y.contrasts = aba.contrasts, 
	y.permutations = aba.permutations)

results <- clean_edgeR(results)

melted_results <- melt_and_stack(results)
```


```{r NADH activity in rifampicin, fig.width=10, fig.height=10}
median_LFC <- dcast(
	melted_results[Pathway == "NADH" & type == "perfect"], 
	unique_name + Pathway ~ condition, 
	value.var = "LFC.adj", 
	fun.aggregate = median)

print(colnames(median_LFC))

to_plot_title <- "Median LogFC in Perfect Guides"

# convert to matrix
LFC_grid <- data.matrix(median_LFC[, c(-1:-2)])

# name rows in matrix
rownames(LFC_grid) <- median_LFC[, paste(unique_name)]

# format genes (rows): Pathway + italic(Gene Name)
genes.formatted <- lapply(
	median_LFC[, .I],
	function(x) { 
		bquote(.(median_LFC[x]$Pathway) ~ italic(.(median_LFC[x]$unique_name)))})

# format conditions (columns)
conditions.formatted <- c(
	bquote(Induction ~ (t[1] - t[0])),
	bquote(Rifampicin ~ (t[1] - t[0])),
	bquote(Rifampicin - Induction ~ (t[1] - t[1])),
	bquote(Induction ~ (t[2] - t[0])),
	bquote(Rifampicin ~ (t[2] - t[0])),
	bquote(Rifampicin - Induction ~ (t[2] - t[2])))

plot_matrix <- LFC_grid

breaks <- generate_breaks(plot_matrix, n = 286, center = T)

to_plot <- pheatmap(
	plot_matrix,
	col = plot_colors,
	breaks = breaks,
	border_color = NA,
	main = to_plot_title,
	cluster_cols = F,
	angle_col = 45,
	clustering_method = "ward.D2",
	clustering_distance_rows = "canberra",
	labels_row = as.expression(genes.formatted),
	labels_col = as.expression(conditions.formatted),
	gaps_col = 3,
	cutree_rows = 3,
	cellheight = 24,
	cellwidth = 48,
	fontsize = 12,
	silent = T)
	
to_plot$gtable$grobs[[1]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[1]]$gp$fontsize <- 18
to_plot$gtable$grobs[[4]]$gp$fontfamily <- "Arial"
to_plot$gtable$grobs[[5]]$gp$fontfamily <- "Arial"
	
print(to_plot)
```



```{r Perform edgeR for Imipenem (high) Samples}
aba.contrasts <- c(
	# "None_0_T1 - None_0_T0",
	# "Imipenem_0.09_T1 - None_0_T0",
	"Imipenem_0.09_T1 - None_0_T1",
	# "None_0_T2 - None_0_T0",
	# "Imipenem_0.09_T2 - None_0_T0",
	"Imipenem_0.09_T2 - None_0_T2") 

results <- extract_edgeR(
	y.fit = aba.fit, 
	y.contrasts = aba.contrasts, 
	y.permutations = aba.permutations)

results <- clean_edgeR(results)

melted_results <- melt_and_stack(results)
```

```{r Cell Wall activity in Imipenem, fig.width=10, fig.height=12}
median_LFC <- melted_results %>% 
	group_by(condition, type, unique_name) %>% 
	mutate(FDR.p = stouffer(FDR)$p, medLFC.adj = median(LFC.adj)) %>% 
	mutate(
    Pathway = NA_character_,
    Pathway = case_when(
        AB030 %in% CL707$AB030 ~ "(Cell Shape/Lipid Glyc.)",
        AB030 %in% CL273$AB030 ~ "(tRNA Aminoacylation)")) %>% 
    filter(!is.na(Pathway)) %>%
    group_by(unique_name, type, condition, Pathway) %>%
    summarise(medLFC.adj = median(LFC.adj), FDR.p = stouffer(FDR)$p) %>% 
    filter(FDR.p < 0.01 & type == "perfect") %>%
    pivot_wider(id_cols = c(unique_name, Pathway), names_from = condition, values_from = medLFC.adj)

print(colnames(median_LFC))

to_plot_title <- "Fitness Changes in Perfect Guides"

# convert to matrix
LFC_grid <- data.matrix(median_LFC[, c(-1:-2)])

# name rows in matrix
rownames(LFC_grid) <- median_LFC %>% pluck("unique_name")

median_LFC <- data.table(median_LFC)

# format genes (rows): Pathway + italic(Gene Name)
genes.formatted <- lapply(
	median_LFC[, .I],
	function(x) {
		bquote(italic(.(median_LFC[x]$unique_name)) ~ .(median_LFC[x]$Pathway))})

# format conditions (columns)
conditions.formatted <- c(
	# bquote(Induction ~ (t[1] - t[0])),
	# bquote(Imipenem ~ (0.09) ~ (t[1] - t[0])),
	bquote(Imipenem ~ (0.09) ~ t[1]),
	# bquote(Induction ~ (t[2] - t[0])),
	# bquote(Imipenem ~ (0.09) ~ (t[2] - t[0])),
	bquote(Imipenem ~ (0.09) ~ t[2]))

plot_matrix <- LFC_grid

breaks <- generate_breaks(plot_matrix, n = 286, center = T)

to_plot <- pheatmap(
	plot_matrix,
	col = plot_colors,
	breaks = breaks,
	border_color = NA,
	main = to_plot_title,
	cluster_cols = F,
	angle_col = 45,
	clustering_method = "ward.D2",
	clustering_distance_rows = "canberra",
	labels_row = as.expression(genes.formatted),
	labels_col = as.expression(conditions.formatted),
	# gaps_col = 3,
	cutree_rows = 3,
	# cellheight = 24,
	cellwidth = 48,
	fontsize = 12,
	silent = T)
	
# to_plot$gtable$grobs[[1]]$gp$fontfamily <- "Courier"
# to_plot$gtable$grobs[[1]]$gp$fontsize <- 18
# to_plot$gtable$grobs[[4]]$gp$fontfamily <- "Courier"
# to_plot$gtable$grobs[[5]]$gp$fontfamily <- "Courier"
	
print(to_plot)
```




```{r Cell Wall activity in Rif and Col, fig.width=10, fig.height=12}

aba.contrasts <- c(
	"Rifampicin_0.34_T2 - None_0_T2",
	"Colistin_0.44_T2 - None_0_T2") 


CL3076 <- fread("../../CL3076_by_name.tsv")
CL849 <- fread("../../CL849_by_name.tsv")

median_melted_results <- fread("../../Results/median_melted_results.tsv.gz")

median_LFC <- median_melted_results %>% 
	filter(condition %in% aba.contrasts) %>% 
	filter(type == "perfect") %>%
	# mutate(Pathway = case_when(
	# 	unique_name %in% CL3076$unique_name ~ "Lipid A",
	# 	unique_name %like% "nuo" ~ "NDH-1",
	# 	TRUE ~ NA_character_)) %>%
	filter(Pathway %in% c("LOS", "NADH")) %>%
	pivot_wider(id_cols = condition, names_from = c(unique_name, Pathway), names_sep = " ", values_from = medLFC)


gene_pathway_labels <-
	median_melted_results %>% 
	filter(condition %in% aba.contrasts) %>% 
	filter(type == "perfect") %>%
	filter(Pathway %in% c("LOS", "NADH")) %>%
	pivot_wider(id_cols = condition, names_from = c(unique_name, Pathway), names_sep = " ", values_from = medLFC) %>%
	colnames() %>% 
	as_tibble() %>% 
	separate(value, c("unique_name", "Pathway"), sep = " ") %>% 
	filter(!is.na(Pathway)) %>%
	as.data.table()

# convert to matrix
LFC_grid <- data.matrix(median_LFC[, c(-1)])

# name rows in matrix
rownames(LFC_grid) <- median_LFC %>% pull(condition)

# format genes (rows): Pathway + italic(Gene Name)
# format genes (rows): Pathway + italic(Gene Name)
genes.formatted <- lapply(
	gene_pathway_labels[, .I],
	function(x) {
		bquote(italic(.(gene_pathway_labels[x]$unique_name)) ~ (.(gene_pathway_labels[x]$Pathway)))})

plot_matrix <- LFC_grid

breaks <- generate_breaks(plot_matrix, n = 286, center = T)

# format conditions (columns)
conditions.formatted <- c(
	bquote(Colistin ~ (T2)),
	bquote(Rifampicin ~ (T2)))

to_plot <- pheatmap(
	plot_matrix,
	col = plot_colors,
	breaks = breaks,
	border_color = NA,
	cluster_row = F,
	angle_col = 315,
	clustering_method = "ward.D2",
	clustering_distance_cols = "canberra",
	labels_row = as.expression(conditions.formatted),
	labels_col = as.expression(genes.formatted),
	# gaps_rows = 2,
	cutree_cols = 2,
	# cellheight = 15,
	# cellwidth = 30,
	fontsize = 12)

	
print(to_plot)
```
